src/ir_Bosch.cpp

```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
+    case stdAc::fanspeed_t::kMax:
+      return kBosch144Fan100;
+    default:
+      return kBosch144FanAuto;
+  }
+}
+
+/// Convert a native mode into its stdAc equivalent.
+/// @param[in] mode The native setting to be converted.
+/// @return The stdAc equivalent of the native setting.
+stdAc::opmode_t IRBosch144AC::toCommonMode(const uint8_t mode) {
+  switch (mode) {
+    case kBosch144Cool: return stdAc::opmode_t::kCool;
+    case kBosch144Heat: return stdAc::opmode_t::kHeat;
+    case kBosch144Dry: return stdAc::opmode_t::kDry;
+    case kBosch144Fan: return stdAc::opmode_t::kFan;
+    default: return stdAc::opmode_t::kAuto;
+  }
+}
+
+/// Convert a native fan speed into its stdAc equivalent.
+/// @param[in] speed The native setting to be converted.
+/// @return The stdAc equivalent of the native setting.
+stdAc::fanspeed_t IRBosch144AC::toCommonFanSpeed(const uint16_t speed) {
+  switch (speed) {
+    case kBosch144Fan100: return stdAc::fanspeed_t::kMax;
+    case kBosch144Fan80: return stdAc::fanspeed_t::kHigh;
+    case kBosch144Fan60: return stdAc::fanspeed_t::kMedium;
+    case kBosch144Fan40: return stdAc::fanspeed_t::kLow;
+    case kBosch144Fan20: return stdAc::fanspeed_t::kMin;
+    default: return stdAc::fanspeed_t::kAuto;
+  }
+}
+
+/// Convert the current internal state into its stdAc::state_t equivalent.
+/// @return The stdAc equivalent of the native settings.
+stdAc::state_t IRBosch144AC::toCommon(void) const {
+  stdAc::state_t result{};
+  result.protocol = decode_type_t::BOSCH144;
+  result.power = getPower();
+  result.mode = toCommonMode(getMode());
+  result.celsius = true;
+  result.degrees = getTemp();
+  result.fanspeed = toCommonFanSpeed(getFan());
+  result.quiet = getQuiet();
+  // Not supported.
+  result.model = -1;
+  result.turbo = false;
+  result.swingv = stdAc::swingv_t::kOff;
+  result.swingh = stdAc::swingh_t::kOff;
+  result.light = false;
+  result.filter = false;
+  result.econo = false;
+  result.clean = false;
+  result.beep = false;
+  result.clock = -1;
+  result.sleep = -1;
+  return result;
+}
+
+/// Convert the current internal state into a human readable string.
+/// @return A human readable string.
+String IRBosch144AC::toString(void) const {
+  uint8_t mode = getMode();
+  uint8_t fan = static_cast<int>(toCommonFanSpeed(getFan()));
+  String result = "";
+  result.reserve(70);  // Reserve some heap for the string to reduce fragging.
+  result += addBoolToString(getPower(), kPowerStr, false);
+  result += addModeToString(mode, kBosch144Auto, kBosch144Cool,
+                            kBosch144Heat, kBosch144Dry, kBosch144Fan);
+  result += addFanToString(fan, static_cast<int>(stdAc::fanspeed_t::kMax),
+                           static_cast<int>(stdAc::fanspeed_t::kMin),
+                           static_cast<int>(stdAc::fanspeed_t::kAuto),
+                           static_cast<int>(stdAc::fanspeed_t::kAuto),
+                           static_cast<int>(stdAc::fanspeed_t::kMedium));
```

Comment:
Rather than use the "common" fan speeds. You're better off using the `kBosch144Fan20` etc values you get directly from `getFan()`.
This way a user is told what value to pass to `setFan()` in the `toString()` output. It also helps with debugging if we get some other "raw" fan speed value.

Reply:
addFanToString() expects 8-bit arguments
kBosch144Fan20 has 9 bits

Reply:
Good point!